### PR TITLE
Add 9.2, drop 9.0 (end of life)

### PR DESCRIPTION
**Pull request by yosifkit**
- https://www.drupal.org/project/drupal/releases/9.2.0
- bump alpine to 3.13 (3.12 is removed from php images)
- add alpine 3.14 for 9.2+
   > Drupal 9.1.x will continue to have security support until December 2021
   > Drupal 8.9.x security coverage ends in November 2021

Fixes https://github.com/docker-library/drupal/issues/196

### CODE REVIEW

File: generate-stackbrew-library.sh [hunk 1]
```diff
@@ -116,6 +114,7 @@ for version; do
 	eval "variants=( $variants )"
 
 	fullVersion="$(jq -r '.[env.version].version' versions.json)"
+	latestAlpineVersion="$(jq -r '.[env.version].variants[] | ltrimstr("fpm-") | select(startswith("alpine")) | ltrimstr("alpine")' versions.json | sort -rV | head -1)"
```
Comment (yosifkit) [inline generate-stackbrew-library.sh:117]: This makes the newest alpine the default for a given drupal version.

File: versions.sh [hunk 2]
```diff
@@ -86,12 +86,39 @@ for version in "${versions[@]}"; do
...
+							[ "alpine3.14", "alpine3.13" ]
+						]
+					end
+					| . as $os
+					| [
+						[ "apache", "fpm" ],
+						[ "fpm" ]
+					] as $variant
+					| keys[]
+					| $variant[.][] as $var | $os[.][] | [ $var, . ] | join("-")
```
Comment (yosifkit) [inline versions.sh:113]: Is this section overkill to generate the variants?

```jq
# 9.2
    "variants": [
      "apache-buster",
      "fpm-buster",
      "fpm-alpine3.14",
      "fpm-alpine3.13"
    ],
# for 10+
    "variants": [
      "apache-bullseye",
      "fpm-bullseye",
      "apache-buster",
      "fpm-buster",
      "fpm-alpine3.14",
      "fpm-alpine3.13"
    ],
```

**Review by yosifkit — COMMENTED** [review]
I might've gone overboard. :ship: :whale:

**Comment by yosifkit** [comment]
```diff
--- ../official-images/library/drupal	2021-06-04 16:58:32.298092190 -0700
+++ /dev/fd/63	2021-06-28 15:32:45.610784653 -0700
@@ -1,53 +1,78 @@
-# this file is generated via https://github.com/docker-library/drupal/blob/1eca4a1de50210dce5d1b443e002e728a838f6b6/generate-stackbrew-library.sh
+# this file is generated via https://github.com/docker-library/drupal/blob/6da7d65e07aa47edf73d632a5b9af9916889bdb2/generate-stackbrew-library.sh
 
 Maintainers: Tianon Gravi <admwiggin@gmail.com> (@tianon),
              Joseph Ferguson <yosifkit@gmail.com> (@yosifkit)
 GitRepo: https://github.com/docker-library/drupal.git
 
-Tags: 9.1.10-php8.0-apache-buster, 9.1-php8.0-apache-buster, 9-php8.0-apache-buster, php8.0-apache-buster, 9.1.10-php8.0-apache, 9.1-php8.0-apache, 9-php8.0-apache, php8.0-apache, 9.1.10-apache-buster, 9.1-apache-buster, 9-apache-buster, apache-buster, 9.1.10-apache, 9.1-apache, 9-apache, apache, 9.1.10, 9.1, 9, latest, 9.1.10-php8.0, 9.1-php8.0, 9-php8.0, php8.0
+Tags: 9.2.0-php8.0-apache-buster, 9.2-php8.0-apache-buster, 9-php8.0-apache-buster, php8.0-apache-buster, 9.2.0-php8.0-apache, 9.2-php8.0-apache, 9-php8.0-apache, php8.0-apache, 9.2.0-apache-buster, 9.2-apache-buster, 9-apache-buster, apache-buster, 9.2.0-apache, 9.2-apache, 9-apache, apache, 9.2.0, 9.2, 9, latest, 9.2.0-php8.0, 9.2-php8.0, 9-php8.0, php8.0
 Architectures: amd64, arm32v7, arm64v8, i386, ppc64le, s390x
-GitCommit: d26d927257e9719ecd764ca64e96263f2374e54e
-Directory: 9.1/php8.0/apache-buster
+GitCommit: 6da7d65e07aa47edf73d632a5b9af9916889bdb2
+Directory: 9.2/php8.0/apache-buster
 
-Tags: 9.1.10-php8.0-fpm-buster, 9.1-php8.0-fpm-buster, 9-php8.0-fpm-buster, php8.0-fpm-buster, 9.1.10-php8.0-fpm, 9.1-php8.0-fpm, 9-php8.0-fpm, php8.0-fpm, 9.1.10-fpm-buster, 9.1-fpm-buster, 9-fpm-buster, fpm-buster, 9.1.10-fpm, 9.1-fpm, 9-fpm, fpm
+Tags: 9.2.0-php8.0-fpm-buster, 9.2-php8.0-fpm-buster, 9-php8.0-fpm-buster, php8.0-fpm-buster, 9.2.0-php8.0-fpm, 9.2-php8.0-fpm, 9-php8.0-fpm, php8.0-fpm, 9.2.0-fpm-buster, 9.2-fpm-buster, 9-fpm-buster, fpm-buster, 9.2.0-fpm, 9.2-fpm, 9-fpm, fpm
 Architectures: amd64, arm32v7, arm64v8, i386, ppc64le, s390x
-GitCommit: d26d927257e9719ecd764ca64e96263f2374e54e
-Directory: 9.1/php8.0/fpm-buster
+GitCommit: 6da7d65e07aa47edf73d632a5b9af9916889bdb2
+Directory: 9.2/php8.0/fpm-buster
 
-Tags: 9.1.10-php8.0-fpm-alpine3.12, 9.1-php8.0-fpm-alpine3.12, 9-php8.0-fpm-alpine3.12, php8.0-fpm-alpine3.12, 9.1.10-php8.0-fpm-alpine, 9.1-php8.0-fpm-alpine, 9-php8.0-fpm-alpine, php8.0-fpm-alpine, 9.1.10-fpm-alpine3.12, 9.1-fpm-alpine3.12, 9-fpm-alpine3.12, fpm-alpine3.12, 9.1.10-fpm-alpine, 9.1-fpm-alpine, 9-fpm-alpine, fpm-alpine
+Tags: 9.2.0-php8.0-fpm-alpine3.14, 9.2-php8.0-fpm-alpine3.14, 9-php8.0-fpm-alpine3.14, php8.0-fpm-alpine3.14, 9.2.0-php8.0-fpm-alpine, 9.2-php8.0-fpm-alpine, 9-php8.0-fpm-alpine, php8.0-fpm-alpine, 9.2.0-fpm-alpine3.14, 9.2-fpm-alpine3.14, 9-fpm-alpine3.14, fpm-alpine3.14, 9.2.0-fpm-alpine, 9.2-fpm-alpine, 9-fpm-alpine, fpm-alpine
 Architectures: amd64, arm32v6, arm32v7, arm64v8, i386, ppc64le, s390x
-GitCommit: d26d927257e9719ecd764ca64e96263f2374e54e
-Directory: 9.1/php8.0/fpm-alpine3.12
+GitCommit: 6da7d65e07aa47edf73d632a5b9af9916889bdb2
+Directory: 9.2/php8.0/fpm-alpine3.14
+
+Tags: 9.2.0-php8.0-fpm-alpine3.13, 9.2-php8.0-fpm-alpine3.13, 9-php8.0-fpm-alpine3.13, php8.0-fpm-alpine3.13, 9.2.0-fpm-alpine3.13, 9.2-fpm-alpine3.13, 9-fpm-alpine3.13, fpm-alpine3.13
+Architectures: amd64, arm32v6, arm32v7, arm64v8, i386, ppc64le, s390x
+GitCommit: 6da7d65e07aa47edf73d632a5b9af9916889bdb2
+Directory: 9.2/php8.0/fpm-alpine3.13
+
+Tags: 9.2.0-php7.4-apache-buster, 9.2-php7.4-apache-buster, 9-php7.4-apache-buster, php7.4-apache-buster, 9.2.0-php7.4-apache, 9.2-php7.4-apache, 9-php7.4-apache, php7.4-apache
+Architectures: amd64, arm32v7, arm64v8, i386, ppc64le, s390x
+GitCommit: 6da7d65e07aa47edf73d632a5b9af9916889bdb2
+Directory: 9.2/php7.4/apache-buster
+
+Tags: 9.2.0-php7.4-fpm-buster, 9.2-php7.4-fpm-buster, 9-php7.4-fpm-buster, php7.4-fpm-buster, 9.2.0-php7.4-fpm, 9.2-php7.4-fpm, 9-php7.4-fpm, php7.4-fpm
+Architectures: amd64, arm32v7, arm64v8, i386, ppc64le, s390x
+GitCommit: 6da7d65e07aa47edf73d632a5b9af9916889bdb2
+Directory: 9.2/php7.4/fpm-buster
+
+Tags: 9.2.0-php7.4-fpm-alpine3.14, 9.2-php7.4-fpm-alpine3.14, 9-php7.4-fpm-alpine3.14, php7.4-fpm-alpine3.14, 9.2.0-php7.4-fpm-alpine, 9.2-php7.4-fpm-alpine, 9-php7.4-fpm-alpine, php7.4-fpm-alpine
+Architectures: amd64, arm32v6, arm32v7, arm64v8, i386, ppc64le, s390x
+GitCommit: 6da7d65e07aa47edf73d632a5b9af9916889bdb2
+Directory: 9.2/php7.4/fpm-alpine3.14
+
+Tags: 9.2.0-php7.4-fpm-alpine3.13, 9.2-php7.4-fpm-alpine3.13, 9-php7.4-fpm-alpine3.13, php7.4-fpm-alpine3.13
+Architectures: amd64, arm32v6, arm32v7, arm64v8, i386, ppc64le, s390x
+GitCommit: 6da7d65e07aa47edf73d632a5b9af9916889bdb2
+Directory: 9.2/php7.4/fpm-alpine3.13
 
-Tags: 9.1.10-php7.4-apache-buster, 9.1-php7.4-apache-buster, 9-php7.4-apache-buster, php7.4-apache-buster, 9.1.10-php7.4-apache, 9.1-php7.4-apache, 9-php7.4-apache, php7.4-apache
+Tags: 9.1.10-php8.0-apache-buster, 9.1-php8.0-apache-buster, 9.1.10-php8.0-apache, 9.1-php8.0-apache, 9.1.10-apache-buster, 9.1-apache-buster, 9.1.10-apache, 9.1-apache, 9.1.10, 9.1, 9.1.10-php8.0, 9.1-php8.0
 Architectures: amd64, arm32v7, arm64v8, i386, ppc64le, s390x
 GitCommit: d26d927257e9719ecd764ca64e96263f2374e54e
-Directory: 9.1/php7.4/apache-buster
+Directory: 9.1/php8.0/apache-buster
 
-Tags: 9.1.10-php7.4-fpm-buster, 9.1-php7.4-fpm-buster, 9-php7.4-fpm-buster, php7.4-fpm-buster, 9.1.10-php7.4-fpm, 9.1-php7.4-fpm, 9-php7.4-fpm, php7.4-fpm
+Tags: 9.1.10-php8.0-fpm-buster, 9.1-php8.0-fpm-buster, 9.1.10-php8.0-fpm, 9.1-php8.0-fpm, 9.1.10-fpm-buster, 9.1-fpm-buster, 9.1.10-fpm, 9.1-fpm
 Architectures: amd64, arm32v7, arm64v8, i386, ppc64le, s390x
 GitCommit: d26d927257e9719ecd764ca64e96263f2374e54e
-Directory: 9.1/php7.4/fpm-buster
+Directory: 9.1/php8.0/fpm-buster
 
-Tags: 9.1.10-php7.4-fpm-alpine3.12, 9.1-php7.4-fpm-alpine3.12, 9-php7.4-fpm-alpine3.12, php7.4-fpm-alpine3.12, 9.1.10-php7.4-fpm-alpine, 9.1-php7.4-fpm-alpine, 9-php7.4-fpm-alpine, php7.4-fpm-alpine
+Tags: 9.1.10-php8.0-fpm-alpine3.13, 9.1-php8.0-fpm-alpine3.13, 9.1.10-php8.0-fpm-alpine, 9.1-php8.0-fpm-alpine, 9.1.10-fpm-alpine3.13, 9.1-fpm-alpine3.13, 9.1.10-fpm-alpine, 9.1-fpm-alpine
 Architectures: amd64, arm32v6, arm32v7, arm64v8, i386, ppc64le, s390x
-GitCommit: d26d927257e9719ecd764ca64e96263f2374e54e
-Directory: 9.1/php7.4/fpm-alpine3.12
+GitCommit: 6da7d65e07aa47edf73d632a5b9af9916889bdb2
+Directory: 9.1/php8.0/fpm-alpine3.13
 
-Tags: 9.0.14-php7.4-apache-buster, 9.0-php7.4-apache-buster, 9.0.14-php7.4-apache, 9.0-php7.4-apache, 9.0.14-apache-buster, 9.0-apache-buster, 9.0.14-apache, 9.0-apache, 9.0.14, 9.0, 9.0.14-php7.4, 9.0-php7.4
+Tags: 9.1.10-php7.4-apache-buster, 9.1-php7.4-apache-buster, 9.1.10-php7.4-apache, 9.1-php7.4-apache
 Architectures: amd64, arm32v7, arm64v8, i386, ppc64le, s390x
-GitCommit: 5d80ce644a0b877e613f7cfb475975b0235cdbcf
-Directory: 9.0/php7.4/apache-buster
+GitCommit: d26d927257e9719ecd764ca64e96263f2374e54e
+Directory: 9.1/php7.4/apache-buster
 
-Tags: 9.0.14-php7.4-fpm-buster, 9.0-php7.4-fpm-buster, 9.0.14-php7.4-fpm, 9.0-php7.4-fpm, 9.0.14-fpm-buster, 9.0-fpm-buster, 9.0.14-fpm, 9.0-fpm
+Tags: 9.1.10-php7.4-fpm-buster, 9.1-php7.4-fpm-buster, 9.1.10-php7.4-fpm, 9.1-php7.4-fpm
 Architectures: amd64, arm32v7, arm64v8, i386, ppc64le, s390x
-GitCommit: 5d80ce644a0b877e613f7cfb475975b0235cdbcf
-Directory: 9.0/php7.4/fpm-buster
+GitCommit: d26d927257e9719ecd764ca64e96263f2374e54e
+Directory: 9.1/php7.4/fpm-buster
 
-Tags: 9.0.14-php7.4-fpm-alpine3.12, 9.0-php7.4-fpm-alpine3.12, 9.0.14-php7.4-fpm-alpine, 9.0-php7.4-fpm-alpine, 9.0.14-fpm-alpine3.12, 9.0-fpm-alpine3.12, 9.0.14-fpm-alpine, 9.0-fpm-alpine
+Tags: 9.1.10-php7.4-fpm-alpine3.13, 9.1-php7.4-fpm-alpine3.13, 9.1.10-php7.4-fpm-alpine, 9.1-php7.4-fpm-alpine
 Architectures: amd64, arm32v6, arm32v7, arm64v8, i386, ppc64le, s390x
-GitCommit: 5d80ce644a0b877e613f7cfb475975b0235cdbcf
-Directory: 9.0/php7.4/fpm-alpine3.12
+GitCommit: 6da7d65e07aa47edf73d632a5b9af9916889bdb2
+Directory: 9.1/php7.4/fpm-alpine3.13
 
 Tags: 8.9.16-php7.4-apache-buster, 8.9-php7.4-apache-buster, 8-php7.4-apache-buster, 8.9.16-php7.4-apache, 8.9-php7.4-apache, 8-php7.4-apache, 8.9.16-apache-buster, 8.9-apache-buster, 8-apache-buster, 8.9.16-apache, 8.9-apache, 8-apache, 8.9.16, 8.9, 8, 8.9.16-php7.4, 8.9-php7.4, 8-php7.4
 Architectures: amd64, arm32v7, arm64v8, i386, ppc64le, s390x
@@ -59,10 +84,10 @@
 GitCommit: 5d80ce644a0b877e613f7cfb475975b0235cdbcf
 Directory: 8.9/php7.4/fpm-buster
 
-Tags: 8.9.16-php7.4-fpm-alpine3.12, 8.9-php7.4-fpm-alpine3.12, 8-php7.4-fpm-alpine3.12, 8.9.16-php7.4-fpm-alpine, 8.9-php7.4-fpm-alpine, 8-php7.4-fpm-alpine, 8.9.16-fpm-alpine3.12, 8.9-fpm-alpine3.12, 8-fpm-alpine3.12, 8.9.16-fpm-alpine, 8.9-fpm-alpine, 8-fpm-alpine
+Tags: 8.9.16-php7.4-fpm-alpine3.13, 8.9-php7.4-fpm-alpine3.13, 8-php7.4-fpm-alpine3.13, 8.9.16-php7.4-fpm-alpine, 8.9-php7.4-fpm-alpine, 8-php7.4-fpm-alpine, 8.9.16-fpm-alpine3.13, 8.9-fpm-alpine3.13, 8-fpm-alpine3.13, 8.9.16-fpm-alpine, 8.9-fpm-alpine, 8-fpm-alpine
 Architectures: amd64, arm32v6, arm32v7, arm64v8, i386, ppc64le, s390x
-GitCommit: 5d80ce644a0b877e613f7cfb475975b0235cdbcf
-Directory: 8.9/php7.4/fpm-alpine3.12
+GitCommit: 6da7d65e07aa47edf73d632a5b9af9916889bdb2
+Directory: 8.9/php7.4/fpm-alpine3.13
 
 Tags: 7.81-php7.4-apache-buster, 7-php7.4-apache-buster, 7.81-php7.4-apache, 7-php7.4-apache, 7.81-apache-buster, 7-apache-buster, 7.81-apache, 7-apache, 7.81, 7, 7.81-php7.4, 7-php7.4
 Architectures: amd64, arm32v5, arm32v7, arm64v8, i386, mips64le, ppc64le, s390x
@@ -74,7 +99,7 @@
 GitCommit: a419d9a8c53e1d27b0f0592dd3a2da99be61df40
 Directory: 7/php7.4/fpm-buster
 
-Tags: 7.81-php7.4-fpm-alpine3.12, 7-php7.4-fpm-alpine3.12, 7.81-php7.4-fpm-alpine, 7-php7.4-fpm-alpine, 7.81-fpm-alpine3.12, 7-fpm-alpine3.12, 7.81-fpm-alpine, 7-fpm-alpine
+Tags: 7.81-php7.4-fpm-alpine3.13, 7-php7.4-fpm-alpine3.13, 7.81-php7.4-fpm-alpine, 7-php7.4-fpm-alpine, 7.81-fpm-alpine3.13, 7-fpm-alpine3.13, 7.81-fpm-alpine, 7-fpm-alpine
 Architectures: amd64, arm32v6, arm32v7, arm64v8, i386, ppc64le, s390x
-GitCommit: a419d9a8c53e1d27b0f0592dd3a2da99be61df40
-Directory: 7/php7.4/fpm-alpine3.12
+GitCommit: 6da7d65e07aa47edf73d632a5b9af9916889bdb2
+Directory: 7/php7.4/fpm-alpine3.13
```

File: versions.sh [hunk 3]
```diff
@@ -86,12 +86,39 @@ for version in "${versions[@]}"; do
 		jq <<<"$json" -c --argjson doc "$doc" '
 			.[env.version] = {
 				version: env.fullVersion,
-				variants: [ "apache-buster", "fpm-buster", "fpm-alpine3.12" ],
+				variants: [
+					# arrays of arrays to control ordering
+					if [ "7", "8.9", "9.1" ] | index(env.version) then
+						[
+							[ "buster" ],
+							[ "alpine3.13" ]
```
Comment (tianon) [inline versions.sh:94]: Do these ones fail on 3.14?

File: versions.sh [hunk 4]
```diff
@@ -86,12 +86,39 @@ for version in "${versions[@]}"; do
...
+						]
+					elif env.version | startswith("9.") then
+						[
+							[ "buster" ],
+							[ "alpine3.14", "alpine3.13" ]
+						]
+					else
+						[
+							[ "bullseye", "buster" ],
```
Comment (tianon) [inline versions.sh:103]: Was this just for testing?  (I don't see any `bullseye` in the `versions.json` and we'll probably want to add it to all once `php` has it, right?)

File: versions.sh (hunk 2 above)
Comment (tianon) [reply versions.sh:113]: It's definitely complicated and a bit hard to follow -- IMO this would be a little more obvious if we flatten the lists above, and then only add the `apache` prefix specially, something like:

```jq
if [ "7", "8.9", "9.1" ] | index(env.version) then
	[
		"buster",
		"alpine3.13"
	]
elif env.version | startswith("9.") then
	[
		"buster",
		#"alpine3.14", "alpine3.13"
		# sometimes I go overkill here with something like:
		(
			"3.14",
			"3.13"
		| "alpine" + .)
	]
else
	[
		"bullseye", "buster",
		"alpine3.14", "alpine3.13"
	]
end
| map(
	if startswith("alpine") then empty else "apache-" + . end,
	"fpm-" + .
)
```

File: versions.sh [hunk 5]
```diff
@@ -86,12 +86,39 @@ for version in "${versions[@]}"; do
...
+				],
 				phpVersions: (
-					if [ "7", "8.9", "9.0"] | index(env.version) then
+					if [ "7", "8.9" ] | index(env.version) then
 						[ "7.4" ]
-					else
+					elif env.version | startswith("9.") then
 						[ "8.0", "7.4" ]
+					else
+						[ "8.0" ]
```
Comment (tianon) [inline versions.sh:121]: Do you think we should add some links in here to why we're doing this? (like that one you found that said 8+ only)

File: versions.sh (hunk 3 above)
Comment (yosifkit) [reply versions.sh:94]: Probably not. I left out `8.9` + `9.1` since they are EOL by the end of the year and `7` just got lumped in even though its EOL is at the end of 2022.

File: versions.sh (hunk 4 above)
Comment (yosifkit) [reply versions.sh:103]: Yeah, was future proofing for Drupal 10 or something.

**Review by tianon — APPROVED** [review]
